### ISSUE TITLE
Enable test execution in browser-wasm

### ISSUE THREAD

**Issue by idg10** [microsoft/testfx#2196]
## Summary

I want to be able to run tests inside a browser-hosted WASM environment. That environment behaves significantly differently from other .NET environments, so it's vital to test thoroughly, but we can't today because there seems to be no support for it.

We were hopeful when the [new test runner](https://devblogs.microsoft.com/dotnet/introducing-ms-test-runner/) was announced that it would address this scenario, since it deals with some other environments that used not to support running tests. But apparently not.

## Background and Motivation

I maintain [Rx.NET (`System.Reactive`)](https://github.com/dotnet/reactive) and we recently had a report of a [problem that occurs in Blazor WASM](https://github.com/dotnet/reactive/issues/2061).

When fixing problems I generally want to add a failing test first to verify that the fix addresses the problem, and to prevent regressions. But I can't do that here because there appears to be no way to run MSTest tests in browser-hosted WASM based .NET code. The problem described occurs only when running in the `browser-wasm` runtime, so for an automated test to reproduce the scenario that triggers this bug, we'd need that test to be able to execute in that runtime.

## Proposed Feature

In an ideal world, we would just be able to add `browser-wasm` (and, since we're talking about WASM also `wasi-wasm`) to the `<RuntimeIdentifiers>` in a test project's `csproj` and for it to automatically run the tests in each listed runtime. But failing that, we'd like some way for `browser-wasm` apps to use the mechanisms that make the new test runner possible.

As far as we can see, at least two things would need to be addressed:

* we'd need some mechanism to capture the output because there's no meaningful standard output in `browser-wasm`
* the mechanism by which configuration is supplied would need to be modified

That second point is the first thing we run into when trying to do this today. Since letting the MSTest tooling generate the app entry point won't work for a `browser-wasm` app, we tried putting code similar to what it generates into our app:

```cs
string[] args = [];
global::Microsoft.Testing.Platform.Builder.ITestApplicationBuilder builder = await 
    global::Microsoft.Testing.Platform.Builder.TestApplication.CreateBuilderAsync(args);
Microsoft.Testing.Platform.MSBuild.TestingPlatformBuilderHook.AddExtensions(builder, args);
Microsoft.Testing.Extensions.Telemetry.TestingPlatformBuilderHook.AddExtensions(builder, args);
Task<ITestApplication> bt = builder.BuildAsync(); // Separating from await because single stepping async in blazor is weird
using (global::Microsoft.Testing.Platform.Builder.ITestApplication app = await bt)
{
    int result = await app.RunAsync();
}
```

This crashes inside `builder.BuildAsync()` call:

```
System.InvalidOperationException: Unexpected state in file '/_/src/Platform/Microsoft.Testing.Platform/Services/CurrentTestApplicationModuleInfo.cs' at line '64'
   at Microsoft.Testing.Platform.Helpers.ApplicationStateGuard.Ensure(Boolean condition, String path, Int32 line) in /_/src/Platform/Microsoft.Testing.Platform/Helpers/ApplicationStateGuard.cs:line 28
   at Microsoft.Testing.Platform.Services.CurrentTestApplicationModuleInfo.GetCurrentTestApplicationFullPath() in /_/src/Platform/Microsoft.Testing.Platform/Services/CurrentTestApplicationModuleInfo.cs:line 64
   at Microsoft.Testing.Platform.Configurations.JsonConfigurationSource.JsonConfigurationProvider.LoadAsync() in /_/src/Platform/Microsoft.Testing.Platform/Configurations/JsonConfigurationProvider.cs:line 32
   at Microsoft.Testing.Platform.Configurations.ConfigurationManager.BuildAsync(IFileLoggerProvider syncFileLoggerProvider) in /_/src/Platform/Microsoft.Testing.Platform/Configurations/ConfigurationManager.cs:line 40
   at Microsoft.Testing.Platform.Hosts.TestHostBuilder.BuildAsync(String[] args, ApplicationLoggingState loggingState, TestApplicationOptions testApplicationOptions, IUnhandledExceptionsHandler unhandledExceptionsHandler, DateTimeOffset createBuilderStart) in /_/src/Platform/Microsoft.Testing.Platform/Hosts/TestHostBuilder.cs:line 134
   at Microsoft.Testing.Platform.Builder.TestApplicationBuilder.BuildAsync() in /_/src/Platform/Microsoft.Testing.Platform/Builder/TestApplicationBuilder.cs:line 118
   at BlazorWasmStandaloneApp.Pages.Home.OnInitializedAsync() in C:\dev\temp\NewMsTestRunner2024Tests\BlazorWasmStandaloneApp\Pages\Home.razor:line 67
   at Microsoft.AspNetCore.Components.ComponentBase.RunInitAndSetParametersAsync()
```

The fundamental problem here is that the MS Test Runner makes the assumption that its JSON configuration file can be found in the filesystem nearby the main assembly. This assumption fails because with `browser-wasm`, the application assembly isn't actually visible on the filesystem. Both `Environment.ProcessPath` and `Assembly.GetEntryAssembly()?.Location` return null because they both represent something that doesn't really exist in `browser-wasm`.

The frustrating thing is that the internal design of the test runner looks like it could easily address these problems. A specialized `IEnvironment` implementation could return a value for the `ProcessPath` even though there isn't really any such file, and then a specialized `IFileSystem` implementation could make the necessary JSON available even though it's not really available in the way the code presumes it will be. These virtualization points look like they would provide a great way to work around `browser-wasm`'s expectation-busting oddity. And console output is also virtualized, which would make it relatively straightforward to capture that output. (This goes on to raise some questions about how you then get those results back in to the IDE or the build output, but it's not really any different from UWP: that's an environment where you are required to launch an actual app with a window just to provide a place to run your tests; in-browser tests have very similar challenges, and a solution exists for UWP.)

But unfortunately, those features are all internal, so we can't use them.

## Alternative Designs

The only alternative seems to be to write our own runner. In principle, nothing stops us from loading up a test adapter component in the same way the real runner does. But this does not seem to be a trivial endeavour.

**Comment by Evangelink** [comment]
Hi @idg10,

Happy to see that the runner seems like a good opportunity for you! 

In all fairness, we did some POC trying to bring support of the runner for "packaged" environments (UWP, WinUI, MAUI...) but we wanted to be able to dedicate enough time to correctly test these environments and so decided to drop it from the scope of the first release. I'll bring it back to the table on our internal discussions so we can make sure to prioritize some time for it.

**Comment by nohwnd** [comment]
@idg10 I will be happy to try this out. Do you have a repro project? That would save me some time 🙂

**Comment by Evangelink** [comment]
I am planning the investigation of this feature for 3.3 milestone, let's see if we can ship a first version by then.

**Comment by Evangelink** [comment]
Postponing to v3.4 as we will be running late for this milestone.

**Comment by Evangelink** [comment]
Hi @idg10 👋

Good news — Microsoft.Testing.Platform (and MSTest on top of it) now boots and runs tests end-to-end on `browser-wasm` (and `wasi-wasm`). The specific crash you hit is fixed: `CurrentTestApplicationModuleInfo` no longer asserts when `Environment.ProcessPath` / `Assembly.GetEntryAssembly().Location` are null in the browser — it now falls back so `BuildAsync()` succeeds. Console output is routed to the browser console via `[JSImport]`, and the single-threaded runtime constraints are handled with inline/synchronous fallbacks.

We''d love for you to try it on your Rx.NET WASM repro. A few pointers:

**How to structure it** — run your tests from a **dedicated `browser-wasm` test app** with its own top-level `Program.Main` (rather than hosting the builder inside a Blazor component, which is what tripped the original attempt). We''ve added a full worked example here: [`samples/BrowserPlayground`](https://github.com/microsoft/testfx/tree/main/samples/BrowserPlayground) — its README walks through `dotnet publish` and running the bundle either **headless under Node** (`node runtests.mjs`) or **in a real browser**.

**Version** — this is on `main` and will ship in the next preview (**MSTest 4.4.0-preview / MTP 2.4.0-preview**). To try it before release, pull our nightly packages from the `test-tools` feed:
`https://pkgs.dev.azure.com/dnceng/public/_packaging/test-tools/nuget/v3/index.json`

**TFM / workload** — the sample targets `net10.0` and needs the `wasm-tools` workload (`dotnet workload install wasm-tools-net10`).

**Known limitations in the browser sandbox** (guarded off by design): TRX report, hang/crash dumps, Azure DevOps report, server mode / the `dotnet test` pipe, `--exit-on-process-exit`, and synchronous file-logger flush. Plain test discovery/execution and console reporting all work.

Would be great to hear how it goes against your actual `browser-wasm` failing scenario — especially anything that doesn''t behave as expected. 🙏

**Comment by idg10** [comment]
Thanks for letting me know.

tl;dr: some initial success, but at some point the tests seem to hang. I've not yet found a way to get it to show me any sort of progress reporting on the tests, making it hard to tell where it has hung. Debugging from Visual Studio doesn't seem to be working usefully: it attaches to the `dotnet.exe` process, and not the NodeJS instance that I believe is actually running the WASM code.

### What went well/ok

I was able to run the `BrowserPlayground` example, but it wasn't obvious to me how to plot a course from there to running our Rx.NET test suite. I've not looked at wasm in Rx since creating this issue back in January 2024, so I'm not exactly fully up to speed with the current state of WASM tooling. I have got something working, but it's entirely possible that it's not the best way.

This repo (`microsoft.testfx`) installs a local .NET SDK, so when I run `dotnet --info` inside my local clone of this repo (having run the build script) it shows `11.0.100-preview.7.26359.110 [D:\microsoft\testfx\.dotnet\sdk]` as one of the installed SDKs. Is that a prerequisite for MTP-on-wasm? My understanding is that WASM support in .NET SDK 10 didn't really come to full fruition, which is why I'm wondering if the .NET 11 SDK preview is necessary.

That said, I was able to use `dotnet new wasmconsole` on a machine with the .NET 10 SDK (and no 11 preview), and then add a reference to `MSTest` `4.4.0-preview.26370.11`, and then with this code:

```cs
ITestApplicationBuilder testApplicationBuilder = await TestApplication.CreateBuilderAsync(args);
testApplicationBuilder.AddMSTest(() => [typeof(MyTestClass).Assembly]);
```

it ran the tests in my test app.

So...I moved on to Rx. I began with this:

```cs
dotnet new wasmconsole -n Tests.System.Reactive.WasmTest
```

I added a reference from that to the existing `Tests.System.Reactive` project. And I added package references to `MSTest` and `MSTest.TestAdapter`. (Initially I tried just `MSTest`, but as far as I could tell, the fact that the existing `Tests.System.Reactive` project already references older version of `MSTest` seemed to be causing a conflict, and adding the explicit reference to the adapter seemed to resolve that.) So my WASM project file looks like this:

```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>net10.0</TargetFramework>
    <RuntimeIdentifier>browser-wasm</RuntimeIdentifier>
    <OutputType>Exe</OutputType>
    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
    <WasmMainJSPath>main.mjs</WasmMainJSPath>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="MSTest" Version="4.4.0-preview.26370.11" />
    <PackageReference Include="MSTest.TestAdapter" Version="4.4.0-preview.26370.11" />
  </ItemGroup>
  <ItemGroup>
    <ProjectReference Include="..\Tests.System.Reactive\Tests.System.Reactive.csproj" />
  </ItemGroup>
</Project>
```

The new WASM project's `Program.cs` looks like this:

```cs
using System;
using System.Runtime.InteropServices;
using System.Runtime.InteropServices.JavaScript;

using Microsoft.Testing.Platform.Builder;
using Microsoft.VisualStudio.TestTools.UnitTesting;

using ReactiveTests.Tests;

Console.WriteLine($"OSVersion: {Environment.OSVersion}");
Console.WriteLine($"IsWasi: {OperatingSystem.IsWasi()}");
Console.WriteLine($"SystemVersion: {RuntimeEnvironment.GetSystemVersion()}");

ITestApplicationBuilder testApplicationBuilder = await TestApplication.CreateBuilderAsync(args);
testApplicationBuilder.AddMSTest(() => [typeof(CountTest).Assembly]);


using ITestApplication testApplication = await testApplicationBuilder.BuildAsync();
return await testApplication.RunAsync();

public partial class MyClass
{
    [JSExport]
    internal static string Greeting()
    {
        var text = $"Hello, World! Greetings from node version: {GetNodeVersion()}";
        return text;
    }

    [JSImport("node.process.version", "main.mjs")]
    internal static partial string GetNodeVersion();
}
```

That `MyClass` is just what the WASM template adds for you. (I didn't remove that, because I was wanted to modify as little as possible of the initial project template just in case I did something daft, causing it all to break for no good reason.) I'm running this with `dotnet run -c Debug` and the initial output starts with the output that the WASM template produces, and then moves on to running my program entry point:

```
Using launch settings from D:\dotnet\reactive\Rx.NET\Source\tests\Tests.System.Reactive.WasmTest\Properties\launchSettings.json...
C:\Users\IanGriffiths\.nuget\packages\microsoft.net.test.sdk\18.8.0\build\net8.0\Microsoft.NET.Test.Sdk.Program.cs(4,41): warning CS7022: The entry point of the program is global code; ignoring 'AutoGeneratedProgram.Main(string[])' entry point.
WasmAppHost --runtime-config D:\dotnet\reactive\Rx.NET\Source\tests\Tests.System.Reactive.WasmTest\bin\Debug\net10.0\browser-wasm\AppBundle\Tests.System.Reactive.WasmTest.runtimeconfig.json
MONO_WASM: Error loading symbol file dotnet.native.js.symbols: {}
Hello, World! Greetings from node version: v24.14.1
OSVersion: Other 1.0.0.0
IsWasi: False
SystemVersion: v10.0.10
MSTest v4.4.0-preview.26370.11 (UTC 20/07/2026)
Tests.System.Reactive.WasmTest.dll (net10.0|browser-wasm)
Test Parallelization enabled for Tests.System.Reactive.dll (Workers: 1, Scope: ClassLevel)
```

and then it starts running tests! 🎉 

I can tell it's doing that because I start seeing test failures. The first one is actually a test we have that ensures we don't accidentally add source files without a copyright header:

```
failed ScanFiles (3ms)
  Microsoft.VisualStudio.TestTools.UnitTesting.MSTestTestNodeException: Test method Tests.System.Reactive.Tests.LicenseHeaderTest.ScanFiles threw exception:
Xunit.Sdk.FalseException: Could not locate sources directory: /
   at Xunit.Assert.False(Nullable`1 condition, String userMessage)
   at Xunit.Assert.False(Boolean condition, String userMessage)
   at Tests.System.Reactive.Tests.LicenseHeaderTest.ScanFiles()
   at System.Reflection.MethodBaseInvoker.InterpretedInvoke_Method(Object obj, IntPtr* args)
   at System.Reflection.MethodBaseInvoker.InvokeWithNoArgs(Object obj, BindingFlags invokeAttr)
```

Unsurprisingly, it can't find the source files from its WASM world. So that's good: we get a test failure in exactly the way you'd expect in this environment.

Next failure:

```
failed BasicsNoSynchronizationContext (13ms)
  Microsoft.VisualStudio.TestTools.UnitTesting.MSTestTestNodeException: Test method Tests.System.Reactive.Tests.TaskLikeSupportTest.BasicsNoSynchronizationContext threw exception:
System.PlatformNotSupportedException: Arg_PlatformNotSupported
   at System.Threading.Thread.ThrowIfNoThreadStart()
   at System.Threading.Thread.Start(Object parameter, Boolean captureContext)
   at System.Threading.Thread.Start(Object parameter)
   at System.Reactive.Concurrency.ConcurrencyAbstractionLayerImpl.StartThread(Action`1 action, Object state)
   at System.Reactive.Concurrency.DefaultScheduler.LongRunning.LongScheduledWorkItem`1[[System.Threading.CancellationToken, System.Private.CoreLib, Version=10.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]]..ctor(CancellationToken state, Action`2 action)
   at System.Reactive.Concurrency.DefaultScheduler.LongRunning.ScheduleLongRunning[CancellationToken](CancellationToken state, Action`2 action)
   at System.Reactive.Linq.ObservableImpl.Delay`1.Base`1.L[[System.Int32, System.Private.CoreLib, Version=10.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.Reactive.Linq.ObservableImpl.Delay`1.Relative[[System.Int32, System.Private.CoreLib, Version=10.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], System.Reactive, Version=7.1.0.0, Culture=neutral, PublicKeyToken=94bc3704cddfc263]].ScheduleDrain()
   at System.Reactive.Linq.ObservableImpl.Delay`1.Relative.L[[System.Int32, System.Private.CoreLib, Version=10.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].RunCore(Relative parent)
   at System.Reactive.Linq.ObservableImpl.Delay`1.Base`1._[[System.Int32, System.Private.CoreLib, Version=10.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.Reactive.Linq.ObservableImpl.Delay`1.Relative[[System.Int32, System.Private.CoreLib, Version=10.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], System.Reactive, Version=7.1.0.0, Culture=neutral, PublicKeyToken=94bc3704cddfc263]].Run(Relative parent)
   at System.Reactive.Linq.ObservableImpl.Delay`1.Relative[[System.Int32, System.Private.CoreLib, Version=10.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].Run(_ sink)
   at System.Reactive.Producer`2.<>c[[System.Int32, System.Private.CoreLib, Version=10.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.Reactive.Linq.ObservableImpl.Delay`1.Base`1._[[System.Int32, System.Private.CoreLib, Version=10.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.Reactive.Linq.ObservableImpl.Delay`1.Relative[[System.Int32, System.Private.CoreLib, Version=10.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], System.Reactive, Version=7.1.0.0, Culture=neutral, PublicKeyToken=94bc3704cddfc263]], System.Reactive, Version=7.1.0.0, Culture=neutral, PublicKeyToken=94bc3704cddfc263]].<SubscribeRaw>b__1_0(ValueTuple`2 tuple)
   at System.Reactive.Concurrency.Scheduler.<>c__75`1[[System.ValueTuple`2[[System.Reactive.Producer`2[[System.Int32, System.Private.CoreLib, Version=10.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.Reactive.Linq.ObservableImpl.Delay`1.Base`1._[[System.Int32, System.Private.CoreLib, Version=10.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.Reactive.Linq.ObservableImpl.Delay`1.Relative[[System.Int32, System.Private.CoreLib, Version=10.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], System.Reactive, Version=7.1.0.0, Culture=neutral, PublicKeyToken=94bc3704cddfc263]], System.Reactive, Version=7.1.0.0, Culture=neutral, PublicKeyToken=94bc3704cddfc263]], System.Reactive, Version=7.1.0.0, Culture=neutral, PublicKeyToken=94bc3704cddfc263],[System.Reactive.Linq.ObservableImpl.Delay`1.Base`1._[[System.Int32, System.Private.CoreLib, Version=10.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.Reactive.Linq.ObservableImpl.Delay`1.Relative[[System.Int32, System.Private.CoreLib, Version=10.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], System.Reactive, Version=7.1.0.0, Culture=neutral, PublicKeyToken=94bc3704cddfc263]], System.Reactive, Version=7.1.0.0, Culture=neutral, PublicKeyToken=94bc3704cddfc263]], System.Private.CoreLib, Version=10.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].<ScheduleAction>b__75_0(IScheduler _, ValueTuple`2 tuple)
   at System.Reactive.Concurrency.CurrentThreadScheduler.Schedule[ValueTuple`2](ValueTuple`2 state, TimeSpan dueTime, Func`3 action)
   at System.Reactive.Concurrency.LocalScheduler.Schedule[ValueTuple`2](ValueTuple`2 state, Func`3 action)
   at System.Reactive.Concurrency.Scheduler.ScheduleAction[ValueTuple`2](IScheduler scheduler, ValueTuple`2 state, Action`1 action)
   at System.Reactive.Producer`2[[System.Int32, System.Private.CoreLib, Version=10.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.Reactive.Linq.ObservableImpl.Delay`1.Base`1._[[System.Int32, System.Private.CoreLib, Version=10.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.Reactive.Linq.ObservableImpl.Delay`1.Relative[[System.Int32, System.Private.CoreLib, Version=10.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], System.Reactive, Version=7.1.0.0, Culture=neutral, PublicKeyToken=94bc3704cddfc263]], System.Reactive, Version=7.1.0.0, Culture=neutral, PublicKeyToken=94bc3704cddfc263]].SubscribeRaw(IObserver`1 observer, Boolean enableSafeguard)
   at System.ObservableExtensions.SubscribeSafe[Int32](IObservable`1 source, IObserver`1 observer)
   at System.Reactive.Linq.QueryLanguage.RunAsync[Int32](IObservable`1 source, CancellationToken cancellationToken)
   at System.Reactive.Linq.QueryLanguage.GetAwaiter[Int32](IObservable`1 source)
   at System.Reactive.Linq.Observable.GetAwaiter[Int32](IObservable`1 source)
   at Tests.System.Reactive.Tests.TaskLikeSupportTest.ManOrBoy_Basics()
   at System.Reactive.PlatformServices.ExceptionServicesImpl.Rethrow(Exception exception)
   at System.Reactive.ExceptionHelpers.Throw(Exception exception)
   at System.Runtime.CompilerServices.TaskObservableMethodBuilder`1.TaskObservable[[System.Int32, System.Private.CoreLib, Version=10.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].GetResult()
   at Tests.System.Reactive.Tests.TaskLikeSupportTest.BasicsNoSynchronizationContext()
   at Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter.Execution.TestMethodInfo.ExecuteInternalAsync(Object[] arguments, CancellationTokenSource timeoutTokenSource)
```

This is exactly the sort of thing I was looking for: I wanted the test suite to tell me which things don't work in WASM, so tests failing with `PlatformNotSupportedException` when trying to use multithreading is precisely what I'd hope for.


### What went less well

The biggest problem is that I can't work out how to diagnose the hang. The debugger only seems to be working for the outer `dotnet.exe` that kicks off NodeJS, and not the actual application running as WASM inside of NodeJS. But I'm wondering if that's just because I'm using .NET SDK 10.0, and Visual Studio 2026—I know WASM support isn't fully baked in .NET 10, so maybe this just doesn't work. Or maybe I need to use VS Code—when I search the web for how to debug WASM on VS, I end up seeing pages talking about VS Code. Any pointers as to how I might get debugging working would be most welcome!

There's also the slightly disconcerting `CS7022` warning. It seems to be related to the fact that the initial `dotnet run` uses the `WasmAppHost`'s entry point. But once that launches NodeJS to run my app properly, it does actually run my `Main` entry point. This seems more likely to be an issue with WASM tooling than anything to do with the test framework though.

**Comment by Evangelink** [comment]
Thank you for the thorough investigation and for testing this against the real Rx.NET suite — this is extremely useful feedback. In particular, the difficulty identifying which test is currently running when execution hangs gives us a clear area where we can improve MTP's browser-WASM diagnostics and progress reporting.

We’ll keep working on the overall experience, including easier execution and tooling integration. One related effort currently in progress is dotnet/sdk#55389, which aims to run WASM test hosts standalone under `dotnet test` with MTP: https://github.com/dotnet/sdk/pull/55389

Thanks again for taking the time to explore this in depth and document what worked and what did not.

**Comment by idg10** [comment]
No problem! Just so I can focus my efforts in the right direction, is it OK for me to continue working with the .NET 10 SDK, or do you think it would be a better idea to move over to the .NET 11 SDK preview if I want to go further with our WASM testing? (We'll move over at some point of course, because we want to be certain that Rx.NET works on .NET 11 when that ships. But we wouldn't usually move over to a preview SDK quite this early in the cycle.)

**Comment by Evangelink** [comment]
I would tend to say net10 should work but I don't know enough APIs changes. I would still expect bigger support on net11. I'll try to find some folks from runtime who could help with answering this question (stay tuned).
